lib/iris/common/metadata.py

```diff
@@ -89,25 +100,433 @@ class BaseMetadata(metaclass=_NamedTupleMeta):
+
+    def _difference_lenient(self, other):
+        """
+        Perform lenient difference of metadata members.
+
+        Args:
+
+        * other (BaseMetadata):
+            The other metadata participating in the lenient difference.
+
+        Returns:
+            A list of difference metadata member values.
+
+        """
+
+        def func(field):
+            left = getattr(self, field)
+            right = getattr(other, field)
+            if field == "units":
+                # Perform "strict" difference for "units".
+                result = None if left == right else (left, right)
+            elif self._is_attributes(field, left, right):
+                result = self._difference_lenient_attributes(left, right)
+            else:
+                # Perform "lenient" difference for members.
+                result = (
+                    (left, right)
+                    if left is not None and right is not None and left != right
+                    else None
+                )
+            return result
+
+        # Note that, we use "_members" not "_fields".
+        return [func(field) for field in BaseMetadata._members]
+
+    @staticmethod
+    def _difference_lenient_attributes(left, right):
+        """Perform lenient difference between the dictionary members."""
+        sleft = set(left.items())
+        sright = set(right.items())
+        # Items in sleft different from sright.
+        dsleft = dict(sleft - sright)
+        # Items in sright different from sleft.
+        dsright = dict(sright - sleft)
+        # Intersection of common item keys with different values.
+        keys = set(dsleft.keys()) & set(dsright.keys())
+        # Keep (in-place) common item keys with different values.
+        [dsleft.pop(key) for key in list(dsleft.keys()) if key not in keys]
+        [dsright.pop(key) for key in list(dsright.keys()) if key not in keys]
+
+        if not bool(dsleft) and not bool(dsright):
+            result = None
+        else:
+            result = (dsleft, dsright)
+
+        return result
+
+    @staticmethod
+    def _difference_strict_attributes(left, right):
+        """Perform strict difference between the dictionary members."""
+        sleft = set(left.items())
+        sright = set(right.items())
+        # Items in sleft different from sright.
+        dsleft = dict(sleft - sright)
+        # Items in sright different from sleft.
+        dsright = dict(sright - sleft)
+
+        if not bool(dsleft) and not bool(dsright):
+            result = None
+        else:
+            result = (dsleft, dsright)
+
+        return result
+
+    @staticmethod
+    def _is_attributes(field, left, right):
+        """Determine whether we have two 'attributes' dictionaries."""
+        return (
+            field == "attributes"
+            and isinstance(left, Mapping)
+            and isinstance(right, Mapping)
+        )
+
+    @lenient_service
+    def combine(self, other, lenient=None):
```

Comment:
Why have you decided to make these public?

Reply:
Sorry, what do you mean? Make what public?

Reply:
I mean that `CubeMetadata.combine`  `CubeMetadata.difference` and  `CubeMetadata.equal` are a part of the public API, so users could happily use them, and expect that in upcoming minor releases their code would still work?

I am possibly being a bit overcautious with what we add to the public API, since this is for a major release, so it would be a long time before we could make changes to anything public.

Reply:
My original intention was always for these methods to be public. 

I've always seen them as offering stable behaviour (i.e., they're not going to fundamentally change from as is now) and that they offer real, tangible benefit and value not only to us for internal `iris` use, but also to users - they now have some rich behaviour to go along with the metadata overhaul that we've banked so far through the common metadata API.

So for me adding them completes this whole common metadata API that we're now offering, and I don't associate any risk with that (certainly no more so that introducing anything else new into `iris`). 

I think that it would be a mistake not to make them public.